File: scripts/solar/common/SolveLuaAsyn.lua
local ____lualib = require("lualib_bundle")
local __TS__Class = ____lualib.__TS__Class
local __TS__SourceMapTraceBack = ____lualib.__TS__SourceMapTraceBack
__TS__SourceMapTraceBack(debug.getinfo(1).short_src, {["8"] = 3,["9"] = 5,["10"] = 6,["11"] = 10,["12"] = 10,["13"] = 10,["15"] = 10,["16"] = 14,["17"] = 16,["18"] = 14,["19"] = 49,["20"] = 50,["21"] = 49,["22"] = 60,["23"] = 61,["24"] = 63,["25"] = 64,["26"] = 65,["27"] = 64,["28"] = 67,["30"] = 68,["31"] = 68,["32"] = 69,["33"] = 70,["36"] = 73,["37"] = 74,["38"] = 68,["42"] = 76,["43"] = 76,["44"] = 77,["45"] = 78,["46"] = 79,["47"] = 76,["50"] = 82,["51"] = 60,["52"] = 86,["53"] = 88,["54"] = 89,["55"] = 88,["56"] = 93,["57"] = 94,["58"] = 93,["59"] = 98,["60"] = 99,["61"] = 98,["62"] = 103,["63"] = 104,["64"] = 103,["65"] = 108,["66"] = 109,["67"] = 108,["68"] = 86,["69"] = 10});
local ____exports = {}
---
-- @noSelfInFile
local CJ = require("jass.common")
local deg = CJ.Rad2Deg(1)
local rad = CJ.Deg2Rad(1)
____exports.default = __TS__Class()
local SolveLuaAsyn = ____exports.default
SolveLuaAsyn.name = "SolveLuaAsyn"
function SolveLuaAsyn.prototype.____constructor(self)
end
function SolveLuaAsyn.init()
    _G.ForGroup = ____exports.default.ForGroup
end
function SolveLuaAsyn.TimerStart(whichTimer, timeout, periodic, handlerFunc)
    CJ.TimerStart(whichTimer, timeout, periodic, handlerFunc)
end
function SolveLuaAsyn.ForGroup(whichGroup, callback)
    local unitHandle = nil
    local oldGetEnumUnit = GetEnumUnit
    _G.GetEnumUnit = function()
        return unitHandle
    end
    local unitArray = {}
    do
        local i = 0
        while i <= 1000000 do
            local unitHandle = FirstOfGroup(whichGroup)
            if not IsHandle(unitHandle) then
                break
            end
            unitArray[#unitArray + 1] = unitHandle
            GroupRemoveUnit(whichGroup, unitHandle)
            i = i + 1
        end
    end
    do
        local i = 0
        while i < #unitArray do
            GroupAddUnit(whichGroup, unitArray[i + 1])
            unitHandle = unitArray[i + 1]
            callback()
            i = i + 1
        end
    end
    _G.GetEnumUnit = oldGetEnumUnit
end
function SolveLuaAsyn.SolveMathAsyn()
    math.sin = function(r)
        return CJ.Sin(r * rad)
    end
    math.cos = function(r)
        return CJ.Cos(r * rad)
    end
    math.tan = function(r)
        return CJ.Tan(r * rad)
    end
    math.asin = function(v)
        return CJ.Asin(v) * deg
    end
    math.acos = function(v)
        return CJ.Acos(v) * deg
    end
end
____exports.default = SolveLuaAsyn
return ____exports

File: src/solar/common/SolveLuaAsyn.ts
/** @noSelfInFile */

const CJ: NoSelf = require('jass.common')
//由于在对战平台 double类型的数学函数 在窗口化时 会丢失精度 所以改用魔兽的float单精浮点数 统一丢掉4字节 防止异步
const deg = CJ.Rad2Deg(1)
const rad = CJ.Deg2Rad(1)

/** @noSelf */
//解决lua异步
export default class SolveLuaAsyn {
    // static tempGroup2: group;


    static init() {
        // SolveLuaAsyn.tempGroup2 = CJ.CreateGroup();
        _G.ForGroup = SolveLuaAsyn.ForGroup;

        /**
         * 命名的函数可能会降低匿名闭包造成的异步几率
         * 即ts里写静态方法 当做回调函数 静态方法里访问静态变量，以减少闭包的对象实例
         * filter 尽量传null 然后在遍历单位组时根据想要的条件if判断 这样就没有闭包到jass了
         */

        //如无暂停计时器窗口这些特殊需求 只是需要简单的几秒后执行函数可使用 BaseUtil.runLater() 内部由lua单个中心计时器回调
        // _G.TimerStart = SolveLuaAsyn.TimerStart;

        /**
         * 优化、减少闭包传入jass
         * 通过一次注册触发 实现触发里再去调用其他需要此事件的触发
         */
        //事件分发
        //_G.TriggerRegisterDeathEvent = TriggerRegisterDeathEvent; //
        /**
         * 自选项
         */
        // @ts-ignore
        // _G.pairs = pairs; //pairs遍历字符串key时可不用 排序后遍历
    }

    /**
     * 命名的函数可能会降低匿名闭包造成的异步几率
     * @constructor
     * @param whichTimer
     * @param timeout
     * @param periodic
     * @param handlerFunc
     */
    //one case todo 这里使用lua模拟的中心计时器 更佳
    static TimerStart(whichTimer: timer, timeout: number, periodic: boolean, handlerFunc: () => void): void {
        CJ.TimerStart(whichTimer, timeout, periodic, handlerFunc)
    }

    //one case
    /**
     * 由于在ForGroup 里不能再调用ForGroup 所以需要2个临时group来中转
     * @param whichGroup
     * @param callback
     * @constructor
     */
    static ForGroup(whichGroup: group, callback: () => void): void {
        let unitHandle: unit = null;
        //重写 GetEnumUnit
        let oldGetEnumUnit = GetEnumUnit;
        _G.GetEnumUnit = function () {
            return unitHandle;
        }
        let unitArray: unit[] = [];
        for (let i = 0; i <= 1000000; i++) {
            let unitHandle = FirstOfGroup(whichGroup)
            if (!IsHandle(unitHandle)) {
                break
            }
            unitArray.push(unitHandle)
            GroupRemoveUnit(whichGroup, unitHandle)
        }
        for (let i = 0; i < unitArray.length; i++) {
            GroupAddUnit(whichGroup, unitArray[i])
            unitHandle = unitArray[i];
            callback()
        }
        //还原 GetEnumUnit
        _G.GetEnumUnit = oldGetEnumUnit;
    };


    static SolveMathAsyn(): void {

        math.sin = function (r: number) {
            return CJ.Sin(r * rad)
        }

        //余弦 
        math.cos = function (r: number) {
            return CJ.Cos(r * rad)
        }

        //正切 
        math.tan = function (r: number) {
            return CJ.Tan(r * rad)
        }

        //反正弦 
        math.asin = function (v: number) {
            return CJ.Asin(v) * deg
        }

        //反余弦
        math.acos = function (v: number) {
            return CJ.Acos(v) * deg
        }

        // //反正切
        // math.atan = function(v1:number, v2:number){
        //  return CJ.atan2(v1, v2) * deg
        // }
    }


}
